Source/Core/VideoBackends/Software/TextureSampler.cpp

```diff
@@ -112,64 +112,109 @@ void Sample(s32 s, s32 t, s32 lod, bool linear, u8 texmap, u8* sample)
   {
-    imageSrc = &texMem[texUnit.texImage1.tmem_even * TMEM_LINE_SIZE];
-    if (texfmt == TextureFormat::RGBA8)
-      imageSrcOdd = &texMem[texUnit.texImage2.tmem_odd * TMEM_LINE_SIZE];
+    state.pointer = &texMem[texUnit.texImage1.tmem_even * TMEM_LINE_SIZE];
+    if (state.texfmt == TextureFormat::RGBA8)
+      state.pointer_odd = &texMem[texUnit.texImage2.tmem_odd * TMEM_LINE_SIZE];
   }
   else
   {
     const u32 imageBase = texUnit.texImage3.image_base << 5;
-    imageSrc = Memory::GetPointer(imageBase);
+    state.pointer = Memory::GetPointer(imageBase);
   }
 
-  int image_width_minus_1 = ti0.width;
-  int image_height_minus_1 = ti0.height;
+  state.width = ti0.width + 1;
+  state.height = ti0.height + 1;
 
   const int tlutAddress = texTlut.tmem_offset << 9;
-  const u8* tlut = &texMem[tlutAddress];
+  state.pointer_tlut = &texMem[tlutAddress];
 
-  // reduce sample location and texture size to mip level
-  // move texture pointer to mip location
-  if (mip)
+  // Precalculate mipmap offsets
+  if (tm0.mipmap_filter != MipMode::None)
   {
-    int mipWidth = image_width_minus_1 + 1;
-    int mipHeight = image_height_minus_1 + 1;
+    int mipWidth = state.width;
+    int mipHeight = state.height;
 
-    const int fmtWidth = TexDecoder_GetBlockWidthInTexels(texfmt);
-    const int fmtHeight = TexDecoder_GetBlockHeightInTexels(texfmt);
-    const int fmtDepth = TexDecoder_GetTexelSizeInNibbles(texfmt);
+    const int fmtWidth = TexDecoder_GetBlockWidthInTexels(state.texfmt);
+    const int fmtHeight = TexDecoder_GetBlockHeightInTexels(state.texfmt);
+    const int fmtDepth = TexDecoder_GetTexelSizeInNibbles(state.texfmt);
 
-    image_width_minus_1 >>= mip;
-    image_height_minus_1 >>= mip;
-    s >>= mip;
-    t >>= mip;
+    u32 mipIdx = 0;
+    u32 offset = 0;
 
-    while (mip)
+    while (mipIdx < maxMips)
     {
       mipWidth = std::max(mipWidth, fmtWidth);
       mipHeight = std::max(mipHeight, fmtHeight);
       const u32 size = (mipWidth * mipHeight * fmtDepth) >> 1;
 
-      imageSrc += size;
+      offset += size;
+      state.mipOffsets[mipIdx++] = offset;
       mipWidth >>= 1;
       mipHeight >>= 1;
-      mip--;
     }
   }
+  else
+  {
+    memset(state.mipOffsets, 0, sizeof(state.mipOffsets));
+  }
+
+  state.manually_managed = texUnit.texImage1.cache_manually_managed;
+  state.wrap_s = tm0.wrap_s;
+  state.wrap_t = tm0.wrap_t;
+}
+void SampleMip(s32 s, s32 t, u32 mip, bool linear, u8 texmap, u8* sample)
+{
+  auto& state = Units[texmap];
+
+  int image_width_minus_1 = state.width - 1;
+  int image_height_minus_1 = state.height - 1;
+
+  u8* imageSrc = state.pointer;
+  u8* imageSrcOdd = state.pointer_odd;
+
+  // reduce sample location and texture size to mip level
+  // move texture pointer to mip location
+  if (mip)
+  {
+    image_width_minus_1 >>= mip;
+    image_height_minus_1 >>= mip;
+    s >>= mip;
+    t >>= mip;
+
+    imageSrc += state.mipOffsets[std::min(maxMips, mip) - 1];
+  }
```

Comment:
Theoretically you could remove the `if` statement here since shifting by 0 does nothing, although you'd need to make `mipOffsets[0]` be 0 (possibly that's why the size is `maxMips + 1` (see earlier comment)?).  I don't know if that'd make much of a difference though.